### PR TITLE
TSAN: fix data race when reading a histogram's sample count

### PR DESCRIPTION
**Pull request by rgs1**
Commit Message: Fix TSAN issue caused from tests when reading histograms.

Additional Description:
This was introduced by #15908. We need to read a histogram's from
the main thread to avoid data races between writes (from the main
thread) and reads (from a worker thread).

Risk Level: Low.
Testing: Unit tests + TSAN should pass.
Docs Changes: NA
Release Notes: NA
Platform Specific Features: NA

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
This should fix the TSAN issue. Happy to move the helper code that does the read from the main thread into
a helper function in test/test_common/utility.h if that's better. 

cc: @antoniovicente @jmarantz

File: test/integration/protocol_integration_test.cc
Comment (antoniovicente) [inline test/integration/protocol_integration_test.cc]: Consider using absl::Notification instead of a mutex and a condvar.

The code here would be something like:
```
uint64_t sample_count = 0;
absl::Notification notification;
test_server_->server().dispatcher().post([&] {
  sample_count = stats.sampleCount();
  notification.Notify();
});

notification.WaitForNotification();
```

File: test/test_common/utility.cc
Comment (antoniovicente) [inline test/test_common/utility.cc]: See comment above about using absl::Notification.

File: test/integration/protocol_integration_test.cc
Comment (rgs1) [reply test/integration/protocol_integration_test.cc]: Updated, thanks.

**Review by antoniovicente — APPROVED** [review]
Thanks!

Waiting for CI before merging.

**Comment by rgs1** [comment]
Ok the ASAN failure seems unrelated:

```
//test/integration:health_check_integration_test                         FAILED in 2 out of 2 in 61.7s
```

I'll merge main...

**Comment by rgs1** [comment]
> Ok the ASAN failure seems unrelated:
> 
> ```
> //test/integration:health_check_integration_test                         FAILED in 2 out of 2 in 61.7s
> ```
> 
> I'll merge main...

Hmm nothing new in main plus it seems possibly transient:

```
test/integration/health_check_integration_test.cc:276: Failure
Value of: clusters_[cluster_idx].host_fake_connection_->waitForNewStream( *dispatcher_, clusters_[cluster_idx].host_stream_)
  Actual: false (Timed out waiting for new stream.)
Expected: true
Stack trace:
  0x160b9b2e: Envoy::(anonymous namespace)::HttpHealthCheckIntegrationTest_SingleEndpointImmediateHealthcheckFailHttp_Test::TestBody()
  0x208969e8: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x20866c13: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x20832340: testing::Test::Run()
  0x20833ad5: testing::TestInfo::Run()
... Google Test internal frames ...

[  FAILED  ] IpHttpVersions/

```

**Comment by rgs1** [comment]
The remaining coverage failure is unrelated, so we can probably just merge:

```
  /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/exe/win32_scm_test/coverage.dat
//test/integration:health_check_integration_test                         FAILED in 38.4s

```

**Comment by rgs1** [comment]
cc: @phlax for visibility on the CI/coverage failure (not sure if it's a global thing)

**Pull request merged**
